### DISCUSSION TITLE
add slice Sum type

### DISCUSSION THREAD

**Discussion by ericsampson** [zeroc-ice/ice#1488] [Ideas]
Sum types are increasingly being adopted by modern languages such as Typescript, Swift, Rust, Kotlin, decent support in recent Java and coming soon to C#, etc

Because they are incredibly useful for robust API modeling. 

Is there a possibility of introducing Sum type support to Slice?

**Comment by bernardnormier** [comment]
Hi Eric,

We call these types "enum with associated values", which is another name for sum types and discriminated unions.

And yes, we want to add enum with associated values to Slice in the near future: it's a useful building block that can't be emulated with the existing constructed types.

The encoding is straightforward: it would be the integral enum value followed by the associated value.

The language mapping is more difficult because Ice supports many languages without direct support for "enum with associated values". 

> coming soon to C#

In C#, we can map an enum with associated values to a record (class) hierarchy, as in:

```slice
enum WebEvent
{
    PageLoad,
    PageUnload,
    KeyPress(byte key),
    Paste(string str),
    Click(long x, long y)
}
```

```csharp
public record WebEvent
{
    public sealed record PageLoad() : WebEvent;
    public sealed record PageUnload() : WebEvent;
    public sealed record KeyPress(byte Key) : WebEvent;
    public sealed record Paste(string Str) : WebEvent;
    public sealed record Click(long X, long Y) : WebEvent;
}
```

Is there an even better solution coming soon?

Best regards,
Bernard

**Comment by bernardnormier** [comment]
Here is the full proposal / discussion to add enum with associated values to Slice: https://github.com/icerpc/slicec/issues/30

We want to implement in Rust before we implement it in C#, since Rust supports this construct natively.
